File: April 2022/April_28.cpp
#define ppii pair<pair<int, int>, int>
class Solution {
public:
    struct cmp {
        constexpr bool operator()(ppii const& a, ppii const& b)
            const noexcept
        {
            return a.second > b.second;
        }
    };
    
    int minimumEffortPath(vector<vector<int>>& heights) {
        int n = heights.size(), m = heights[0].size();
        bool vis[n][m];
        memset(vis, false, sizeof(vis));
        
        priority_queue<ppii, vector<ppii>, cmp> pq;
        pq.push({{0, 0}, 0});
        
        while(!pq.empty()){
            ppii rem = pq.top(); pq.pop();
            int i = rem.first.first, j = rem.first.second;
            if(vis[i][j])
                continue;
            vis[i][j] = true;
            if(i == n - 1 && j == m - 1)
                return rem.second;
            if(i > 0)
                pq.push({{i - 1, j}, max(rem.second, abs(heights[i - 1][j] - heights[i][j]))});
            if(j > 0)
                pq.push({{i, j - 1}, max(rem.second, abs(heights[i][j - 1] - heights[i][j]))});
            if(i + 1 < n)
                pq.push({{i + 1, j}, max(rem.second, abs(heights[i + 1][j] - heights[i][j]))});
            if(j + 1 < m)
                pq.push({{i, j + 1}, max(rem.second, abs(heights[i][j + 1] - heights[i][j]))});
        }
        
        return 0;
    }
};

File: April 2022/April_28.java
class Solution {
    public int minimumEffortPath(int[][] heights) {
        int n = heights.length, m = heights[0].length;
        boolean[][] vis = new boolean[n][m];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.add(new int[]{0, 0, 0});
        while(!pq.isEmpty()){
            int[] rem = pq.remove();
            if(vis[rem[0]][rem[1]])
                continue;
            vis[rem[0]][rem[1]] = true;
            if(rem[0] == n - 1 && rem[1] == m - 1)
                return rem[2];
            
            if(rem[0] > 0)
                pq.add(new int[]{rem[0] - 1, rem[1], Math.max(rem[2], Math.abs(heights[rem[0]][rem[1]] - heights[rem[0] - 1][rem[1]]))});
            if(rem[1] > 0)
                pq.add(new int[]{rem[0], rem[1] - 1, Math.max(rem[2], Math.abs(heights[rem[0]][rem[1]] - heights[rem[0]][rem[1] - 1]))});
            if(rem[0] + 1 < n)
                pq.add(new int[]{rem[0] + 1, rem[1], Math.max(rem[2], Math.abs(heights[rem[0]][rem[1]] - heights[rem[0] + 1][rem[1]]))});
            if(rem[1] + 1 < m)
                pq.add(new int[]{rem[0], rem[1] + 1, Math.max(rem[2], Math.abs(heights[rem[0]][rem[1]] - heights[rem[0]][rem[1] + 1]))});
        }
        return 0;
    }
}
